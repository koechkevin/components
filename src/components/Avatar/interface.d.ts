import { AvatarProps } from 'antd/lib/avatar';
import { CSSProperties, ReactNode } from 'react';

enum ChatStatus {
  online = 'online',
  offline = 'offline',
}

export interface Props extends AvatarProps {
  src?: string;
  dot?: boolean;
  status?: ChatStatus | string;
  size?: number | 'xsmall' | 'small' | 'medium' | 'large' | 'default';
  color?: string;
}
