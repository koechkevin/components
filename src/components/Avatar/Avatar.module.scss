@import '../../styles/global';

.avatar {
  &:global(.ant-avatar-square) {
    border-radius: 8px;
  }

  &:global(.ant-avatar img) {
    background-color: $color-basic-light-gray;
  }
}
