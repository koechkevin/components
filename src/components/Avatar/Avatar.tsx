import { Avatar as AntdAvatar } from 'antd';
import React, { FC } from 'react';

import styles from './Avatar.module.scss';

import { Props } from './interface.d';

export const Avatar: FC<Props> = (props) => {
  const { children, className, icon, src, shape, size, style, color } = props;
  const avatarSizes = {
    small: 32,
    medium: 40,
    large: 80,
    default: 'default',
  };
  const avatarSize = size && (avatarSizes[size] || size);
  const placeholderIcon = src ? icon : 'user';
  const classNames: string = [styles.avatar, className ? className : ''].join(' ');
  const avatarColor = color && !src ? { backgroundColor: color } : {};

  return (
    <AntdAvatar
      src={src}
      shape={shape}
      size={avatarSize}
      className={classNames}
      style={{ minWidth: avatarSize, ...avatarColor, ...style }}
      icon={!children ? placeholderIcon : ''}
    >
      {children}
    </AntdAvatar>
  );
};

Avatar.defaultProps = {
  size: 'default',
};

export default Avatar;
