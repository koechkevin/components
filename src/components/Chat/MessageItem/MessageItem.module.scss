@import '../../../styles/message';

.messageItem {
  @extend .message;

  :global(.ant-list-item-meta-title) {
    margin-bottom: 0;
  }
}

.deletedWrapper {
  border-radius: 0px;
  padding-left: 85px;
  background-color: $color-gray-light;

  .deletedMessage {
    color: $color-base-dark 60%;
    font-style: italic;
    font-size: 14px;
    justify-content: start;
  }

  .banIcon {
    width: 16px;
    height: 16px;
    margin-right: 4px;
  }
}

.emptyAvatar {
  padding-left: 40px;
}

.replies {
  padding: 8px;
  cursor: pointer;
  overflow: hidden;
  border-radius: 4px;
  border: 1px solid transparent;

  &:hover {
    border: 1px solid $color-base-gray;

    .timestamp {
      display: none;
    }

    .viewThreadText {
      display: inline;
    }

    .threadChevron {
      display: inline;
    }
  }

  .replyInfo {
    flex: 1 auto;
    overflow: hidden;
    text-overflow: ellipsis;
    white-space: nowrap;
  }

  .viewThreadText {
    display: none;
    color: $color-base-gray-dark;
  }

  .threadChevron {
    float: right;
    padding-right: 10px;
    vertical-align: middle;
    display: none;
  }

  span {
    font-size: 12px;
    line-height: 24px;
  }

  .blue {
    padding: 0 4px;
    font-size: 12px;
    font-weight: bold;
    color: $color-primary;
  }

  .gray {
    padding: 0 4px;
    color: $color-base-gray-dark;
  }

  .divider {
    color: $color-base-gray-dark;
  }

  :global {
    .ant-avatar {
      margin-right: 4px;
    }
  }
}
