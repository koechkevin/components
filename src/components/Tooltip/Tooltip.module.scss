@import "../../styles/global";

.tooltip {
  :global(.ant-tooltip-inner) {
    font-size: 12px;
    background-color: $color-base-dark;
  }
}

.light {
  :global {
    .ant-tooltip-inner {
      font-size: 12px;
      color: $color-base-dark;
      background-color: $color-white;
      box-shadow: 0 2px 4px rgba($color: $color-black, $alpha: 0.16);
    }

    .ant-tooltip-arrow::before {
      background-color: $color-white;
    }
  }
}