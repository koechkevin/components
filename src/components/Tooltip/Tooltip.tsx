import { Tooltip as AntdTooltip } from 'antd';
import { AbstractTooltipProps, RenderFunction } from 'antd/lib/tooltip';
import React, { FC, ReactNode } from 'react';

import styles from './Tooltip.module.scss';

interface Props extends AbstractTooltipProps {
  delay?: number;
  light?: boolean;
  title?: RenderFunction | ReactNode;
}

export const Tooltip: FC<Props> = (props) => {
  const { title, light, children, delay, ...restProps } = props;
  const isTouchDevice = () => 'ontouchstart' in window;
  const overlay = light ? styles.light : styles.tooltip;

  return (
    <AntdTooltip
      mouseEnterDelay={delay}
      mouseLeaveDelay={delay}
      title={!isTouchDevice() && title}
      overlayClassName={overlay}
      {...restProps}
    >
      {children}
    </AntdTooltip>
  );
};

Tooltip.defaultProps = {
  delay: 0.0001,
};

export default Tooltip;
