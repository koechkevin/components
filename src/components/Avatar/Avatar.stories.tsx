import { number, text } from '@storybook/addon-knobs';
import { storiesOf } from '@storybook/react';
import { Col, Row, Typography } from 'antd';
import React from 'react';

import { Center } from '../Center';
import { ListItem } from '../SideBar/CompanyModal';
import Avatar from './Avatar';

import logo from '../../icons/logo.svg';
import mockLogo from '../../icons/mocks/company.png';

const src = 'https://s3.amazonaws.com/uifaces/faces/twitter/bowbrick/128.jpg';
const size = number('size', 200);
const initials = text('initials', 'AA');
const { Text } = Typography;
const companyList: ListItem[] = [
  { companyId: '123', avatarUrl: logo, name: 'Airbus', count: 4, avatarColor: '#fb6d77' },
  { companyId: '1234', avatarUrl: mockLogo, name: 'Pepsi', count: 21, avatarColor: '#46a094' },
  { companyId: '12345', avatarUrl: '', name: 'Airbus', avatarColor: '#5d85ce' },
];

storiesOf('Avatar', module)
  .addDecorator((story) => <Center>{story()}</Center>)
  .add('default', () => <Avatar />, { notes: 'Avatar default styles' })
  .add(
    'with circle shape and image',
    () => (
      <Row>
        <Row gutter={36}>
          <Col span={8}>
            <Text strong>24</Text> xsmall{' '}
          </Col>
          <Col span={8}>
            <Text strong>40</Text> small{' '}
          </Col>
          <Col span={8}>
            <Text strong>80</Text> medium{' '}
          </Col>
        </Row>
        <Row gutter={36}>
          <Col span={8}>
            <Avatar src={src} size="small" />
          </Col>
          <Col span={8}>
            <Avatar src={src} size="medium" />
          </Col>
          <Col span={8}>
            <Avatar src={src} size="large" />
          </Col>
        </Row>
      </Row>
    ),
    {
      notes: 'Avatar with Circle shape and image',
    },
  )
  .add(
    'with square shape and image',
    () => (
      <Row>
        <Row gutter={36}>
          <Col span={8}>
            <Text strong>24</Text> xsmall
          </Col>
          <Col span={8}>
            <Text strong>40</Text> small
          </Col>
          <Col span={8}>
            <Text strong>80</Text> medium
          </Col>
        </Row>
        <Row gutter={36}>
          <Col span={8}>
            <Avatar src={src} shape="square" size="small" />
          </Col>
          <Col span={8}>
            <Avatar src={src} shape="square" size="medium" />
          </Col>
          <Col span={8}>
            <Avatar src={src} shape="square" size="large" />
          </Col>
        </Row>
      </Row>
    ),
    {
      notes: 'Avatar with square shape and image',
    },
  )
  .add(
    'with square shape and initials',
    () => (
      <Row>
        <Row>
          <Col span={12}>
            <Text strong>32</Text> small
          </Col>
          <Col span={12}>
            <Text strong>40</Text> large
          </Col>
        </Row>
        <Row gutter={36}>
          <Col span={12}>
            <Avatar size="small" shape="square" color="#46a094" style={{ fontWeight: 'bold', fontSize: 10 }}>
              {initials}
            </Avatar>
          </Col>
          <Col span={12}>
            <Avatar size="medium" shape="square" color="#8260da" style={{ fontWeight: 'bold', fontSize: 20 }}>
              {initials}
            </Avatar>
          </Col>
        </Row>
      </Row>
    ),
    {
      notes: 'Avatar with Square shape and initials',
    },
  )
  .add(
    'with circle shape and initials',
    () => (
      <Row>
        <Row gutter={36}>
          <Col span={8}>
            <Text strong>32</Text> small
          </Col>
          <Col span={8}>
            <Text strong>40</Text> medium
          </Col>
          <Col span={8}>
            <Text strong>80</Text> large
          </Col>
        </Row>
        <Row gutter={36}>
          <Col span={8}>
            <Avatar size="small" shape="circle" color="#46a094" style={{ fontWeight: 'bold', fontSize: 10 }}>
              {initials}
            </Avatar>
          </Col>
          <Col span={8}>
            <Avatar size="medium" shape="circle" color="#8260da" style={{ fontWeight: 'bold', fontSize: 20 }}>
              {initials}
            </Avatar>
          </Col>
          <Col span={8}>
            <Avatar size="large" shape="circle" color="#8260da" style={{ fontWeight: 'bold', fontSize: 20 }}>
              {initials}
            </Avatar>
          </Col>
        </Row>
      </Row>
    ),
    {
      notes: 'Avatar with Square shape and initials',
    },
  )
  .add('with custom size', () => <Avatar src={src} size={size} />, {
    notes: 'Avatar with custom size',
  })
  .add('with an icon', () => <Avatar icon="user" />, {
    notes: 'Avatar with an icon',
  })
  .add('company logo', () => (
    <Row gutter={32}>
      {companyList.map((company) => (
        <Col span={8} key={company.companyId}>
          <Avatar src={company.avatarUrl} size="medium" shape="square" color={company.avatarColor}>
            {company.name[0]}
          </Avatar>
        </Col>
      ))}
    </Row>
  ));
