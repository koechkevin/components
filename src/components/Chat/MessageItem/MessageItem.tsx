import { faBan } from '@fortawesome/free-solid-svg-icons';
import { faChevronRight } from '@fortawesome/pro-light-svg-icons';
import { FontAwesomeIcon } from '@fortawesome/react-fontawesome';
import { List, Row } from 'antd';
import React, { FC, useEffect, useState } from 'react';

import { FileModel, MessageAuthor, MessageItemProps } from '../../../typings/Message';
import {
  FileExtension,
  getNameInitials,
  ProductType,
  shortTimeStampFormat,
  timestampFormat,
  urlConverter,
} from '../../../utils';
import { Avatar } from '../../Avatar';
import { FileItem } from '../../FileUpload';
import { Tooltip } from '../../Tooltip';
import ForwardedMessage from '../ForwardedMessage/ForwardedMessage';
import { Signature } from './../../Signature';

import styles from './MessageItem.module.scss';

const { Item } = List;
let element: any;

const repliesRender = (replies: number): string => {
  if (replies > 1) {
    return `${replies} Replies`;
  }

  if (replies === 1) {
    return `1 Reply`;
  }

  return '';
};

const replyPeopleRender = (authors: MessageAuthor[]): string => {
  const count: number = authors.length - 3;
  return `+ ${count} ${count > 1 ? 'People' : 'Person'}`;
};

interface MessageProps extends MessageItemProps {
  menuShown?: boolean;
  poppedMenu: React.ReactNode;
  onCancel?: (file: FileModel) => void;
  onDownload?: (file: FileModel) => void;
  startThread: (messageId: string) => void;
}

const MessageItem: FC<MessageProps> = (props) => {
  const {
    id,
    text,
    file,
    author,
    onCancel,
    createdAt,
    updatedAt,
    forwarding,
    threadInfo,
    isModified,
    isDeleted,
    parentId,
    menuShown,
    poppedMenu,
    startThread,
    onDownload,
  } = props;
  const [shown, setShown] = useState(false);

  const handleReply = () => {
    if (startThread && id) {
      startThread(id);
    }
  };
  const addTouchStart = (event: any) => {
    event.preventDefault();
    const touch = event.touches[0];
    element = document.elementFromPoint(touch.pageX, touch.pageY);
  };

  const addTouchMove = (event: any) => {
    event.preventDefault();
    const touch = event.touches[0];
    if (element !== document.elementFromPoint(touch.pageX, touch.pageY)) {
      setShown(false);
    }
  };

  useEffect(() => {
    document.addEventListener('touchstart', addTouchStart, false);
    document.addEventListener('touchmove', addTouchMove, false);

    return () => {
      document.removeEventListener('touchstart', addTouchStart, false);
      document.removeEventListener('touchmove', addTouchMove, false);
    };
  }, []);

  return (
    <>
      {isDeleted && (
        <Row className={styles.deletedWrapper}>
          <Item className={styles.deletedMessage}>
            <span>
              <FontAwesomeIcon icon={faBan} className={styles.banIcon} />
            </span>
            <span>This message was deleted</span>
          </Item>
        </Row>
      )}
      <Item
        key={id}
        className={styles.messageItem}
        onMouseEnter={() => setShown(true)}
        onMouseLeave={() => setShown(false)}
        onTouchStart={() => setShown(true)}
      >
        {/** popover menu */}
        {menuShown && shown && id && poppedMenu}
        {/** message item detail */}
        <Item.Meta
          avatar={
            <>
              {isDeleted ? (
                <span className={styles.emptyAvatar} />
              ) : (
                <Avatar
                  size={40}
                  src={author && author.avatar}
                  status={author && author.chatStatus}
                  style={{ background: author && author.avatarColor, fontWeight: 'bold' }}
                  shape={author && author.productId === ProductType.Candidate ? 'square' : 'circle'}
                >
                  {author && getNameInitials(author.name)}
                </Avatar>
              )}
            </>
          }
          title={
            <Row>
              {!isDeleted && (
                <>
                  <Row>
                    <span className={styles.name}>{author && author.name}</span>
                    <span className={styles.created}>
                      {parentId
                        ? createdAt && timestampFormat(new Date(createdAt))
                        : createdAt && shortTimeStampFormat(new Date(createdAt))}
                    </span>
                  </Row>
                  {author && author.signature && (
                    <Row>
                      <Signature
                        status={'Owner'}
                        visibility={'visible'}
                        label={author && author.label}
                        name={author && author.signature}
                      />
                    </Row>
                  )}
                  <Row className={styles.text}>
                    <span dangerouslySetInnerHTML={{ __html: urlConverter(text) || '' }} />
                    {isModified && (
                      <Tooltip placement="top" title={updatedAt && `${shortTimeStampFormat(new Date(updatedAt))}`}>
                        <span className={styles.tag}>(Edited)</span>
                      </Tooltip>
                    )}
                  </Row>
                  {file && (
                    <FileItem
                      {...file}
                      downloadable={!!file.fileId}
                      style={{ margin: '8px 0' }}
                      name={file.filename || file.name}
                      extension={file.extension as FileExtension}
                      onCancel={() => onCancel && onCancel(file)}
                      onDownload={() => onDownload && onDownload(file)}
                    />
                  )}
                </>
              )}
            </Row>
          }
          description={
            <>
              {forwarding && <ForwardedMessage {...forwarding} />}
              {threadInfo && threadInfo.authors && (
                <Row
                  type="flex"
                  align="middle"
                  justify="space-between"
                  onClick={handleReply}
                  className={styles.replies}
                >
                  <Row className={styles.replyInfo}>
                    {threadInfo.authors.slice(0, 3).map((author: MessageAuthor, idx: number) => {
                      return (
                        <Avatar
                          size={24}
                          key={idx}
                          src={author.avatar}
                          style={{
                            fontSize: 12,
                            fontWeight: 'bold',
                            backgroundColor: author.avatarColor,
                          }}
                          shape={author.productId === 'candidate-portal' ? 'square' : 'circle'}
                        >
                          {getNameInitials(author.name)}
                        </Avatar>
                      );
                    })}
                    {threadInfo.authors.length > 3 && (
                      <span className={styles.blue}>{replyPeopleRender(threadInfo.authors)}</span>
                    )}
                    {threadInfo.authors.length > 3 && <span className={styles.divider}>-</span>}
                    <span className={`${threadInfo.authors.length > 3 ? styles.gray : styles.blue}`}>
                      {repliesRender(threadInfo.count)}
                    </span>
                    {threadInfo.authors.length > 3 && <span className={styles.divider}>-</span>}
                    <span className={`${styles.gray} ${styles.timestamp}`}>
                      Last reply {threadInfo.lastMessageAt && shortTimeStampFormat(new Date(threadInfo.lastMessageAt))}
                    </span>
                    <span className={styles.viewThreadText}> View Thread </span>
                  </Row>
                  <span className={styles.threadChevron}>
                    <FontAwesomeIcon icon={faChevronRight} />
                  </span>
                </Row>
              )}
            </>
          }
        />
      </Item>
    </>
  );
};

export default MessageItem;
