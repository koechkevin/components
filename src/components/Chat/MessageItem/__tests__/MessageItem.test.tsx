import { mount } from 'enzyme';
import React from 'react';

import { shortTimeStampFormat, timestampFormat } from '../../../../utils';
import { Avatar } from '../../../Avatar';
import { Tooltip } from '../../../Tooltip';
import MessageItem from '../MessageItem';

describe('Tests for MessageItem component', () => {
  const props = {
    id: 'sampleMessageId',
    createdAt: '2019-10-08 06:22:09',
    updatedAt: '2020-10-08 06:22:09',
    author: {
      id: 'user-fake-id',
      name: 'Jane Doe',
      avatarColor: '#4224',
      productId: 'company-portal',
      chatStatus: 'sample chat status',
    },
    text: 'sample text',
    poppedMenu: <div>Mock Node</div>,
    menuShown: true,
    startThread: jest.fn(),
    forwarding: null,
  };

  const wrapper = mount(<MessageItem {...props} />);

  it('should render message item without crushing', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('should render avatar for company portal', () => {
    const avatar = wrapper.find(Avatar).first();

    expect(avatar.props().size).toEqual(40);
    expect(avatar.props().shape).toEqual('circle');
    expect(avatar.props().status).toEqual(props.author.chatStatus);
  });

  it('should render correct timestamp', () => {
    expect(wrapper.contains(shortTimeStampFormat(new Date(props.createdAt)))).toEqual(true);
  });

  it('should render tooltip with timestamp title and is modified property if modified', () => {
    const updatedProps = {
      ...props,
      isModified: true,
    };

    const updatedWrapper = mount(<MessageItem {...updatedProps} />);
    const toolTip = updatedWrapper.find(Tooltip).first();

    expect(toolTip.props().title).toEqual(shortTimeStampFormat(new Date(updatedProps.updatedAt)));
    expect(toolTip.contains('(Edited)')).toEqual(true);
  });

  it('should render actual date and time if message is in a thread', () => {
    const updatedProps = {
      ...props,

      parentId: '5dcd77257d216a0022d2e803',
    };
    const updatedWrapper = mount(<MessageItem {...updatedProps} />);
    expect(updatedWrapper.contains(timestampFormat(new Date(props.createdAt)))).toEqual(true);
  });

  it('should render deleted message text if main message is deleted', () => {
    const updatedProps = {
      ...props,
      threadInfo: {
        lastMessageAt: '2020-10-08 06:22:09',
        count: 1,
        authors: [
          {
            id: 'mock user id',
            name: 'jane Doe',
            status: 'mock status',
            avatarColor: '#5D85CE',
            chatStatus: 'online',
            productId: 'candidate-portal',
            createdAt: '2019-10-08 06:22:09',
          },
        ],
      },
      isDeleted: true,
    };

    const updatedWrapper = mount(<MessageItem {...updatedProps} />);
    expect(updatedWrapper.find('.deletedMessage').exists()).toEqual(true);
    expect(updatedWrapper.contains('This message was deleted')).toBeTruthy();
  });

  it('should test for description prop', () => {
    const updatedProps = {
      ...props,
      threadInfo: {
        lastMessageAt: '2020-10-08 06:22:09',
        count: 1,
        authors: [
          {
            id: 'mock user id',
            name: 'jane Doe',
            status: 'mock status',
            productId: 'candidate-portal',
            createdAt: '2019-10-08 06:22:09',
            avatarColor: '#dfc475',
            chatStatus: 'offline',
          },
        ],
      },
    };

    const updatedWrapper = mount(<MessageItem {...updatedProps} />);
    expect(updatedWrapper.find('span.blue').exists()).toEqual(true);
    expect(
      updatedWrapper.contains(`Last reply ${shortTimeStampFormat(new Date(updatedProps.threadInfo.lastMessageAt))}`),
    );
    expect(wrapper.contains('View Thread'));
  });
});
