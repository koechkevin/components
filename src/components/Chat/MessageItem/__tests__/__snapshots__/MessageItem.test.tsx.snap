// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tests for MessageItem component should render message item without crushing 1`] = `ReactWrapper {}`;
